### DISCUSSION TITLE
네이밍 컨벤션 관련 질문

### DISCUSSION THREAD

**Discussion by SeungrokYoon** [wanted-pre-onboarding-team-12th-7/pre-onboarding-12th-1-7#9] [Announcements]
같이 라이브쉐어로 페어코딩을 하면서 우리가 네이밍 스타일이 많이 다르다는 것을 느꼈습니다. 

저희가 회의때 네이밍 컨벤션에 대해 세세하게 정하지 않아도 괜찮다고 생각했는데, 그 필요성을 느끼게 되어 이슈를 작성하게 되었습니다. 

스타일이 크게 다르다고 느낀 점은 타입과 인터페이스 관련 부분이었습니다.

## 팀원들의 네이밍 습관

가령, Button 컴포넌트의 prop에 대한 인터페이스는 

```
interface IButtonProps {
...
}

```
처럼 I를 붙이는 헝가리안 표기법(타입임을 표시하는 `I`또는 `T` + `변수명` + `"Props"접미사`)을 사용하거나, 

```
interface ButtonProps {
...
}

```

처럼 `컴포넌트명` + `"Props"접미사` 를 사용하는 경우가 있었으며, 

단순히 

```
interfaceProps {
...
}

```
로 사용하는 경우가 있었습니다. 

## 제안

타입스크립트의 강력한 이점 중 하나는 `타입을 재사용` 할 수 있다는 것입니다. 

그렇기에 아래와 같은 컨벤션을 제안하고 싶습니다.

순수 인터페이스는 아래와같이 접두사, 접미사 없이 대문자로 시작하는 단순 변수명으로 사용하고,
```
interface Sample {
  target:string; ...
}

type Sample {
  target:string; ...
}
```

컴포넌트 props는 자주 재사용되니, 컴포넌트명 + "Props" 접미사를 붙여서 명시해주면 좋을 것 같습니다.
```
interface SampleButtonProps {
  target:string; ...
}
```

## 그래서...
그 외에 애매하다 싶은 네이밍은 이 곳에서 질문을 남기고 논의하면 좋을 것 같습니다!

**Comment by salmontaker** [comment]
저는 동의합니다! 단순 파스칼 케이스로 작성하는것이 더 보기좋고 깔끔한것 같아요.

**Comment by wisdomin121** [comment]
저도 동의합니다 ~!

항상 써왔던 네이밍이 헝가리안 표기법인지 몰랐네요 ㅎㅎ..
이미 올려뒀던 코드들도 파스칼 케이스로 바꾸도록 하겠습니다 !

**Comment by 5unk3n** [comment]
저도 동의합니다.

그리고 저는 주로 객체에는 interface를 사용하고 나머지 타입은 type을 사용하는데 다른 분들은 어떤 방식으로 사용하는지도 궁금합니다!

**Reply by SeungrokYoon** [reply]
저는 컴포넌트 프롭에서는 interface를 사용하는 편이 더 좋아보입니다. 
```js
interface ButtonProps extends HTMLAttribute<HTMLButtonElement> {
   ...
}
```

처럼, 다른 타입을 extends 하는 경우가 잦은 것 같아요~ 다른 컴포넌트에서 호출되어 사용되는 경우도 잦구요.

**Reply by salmontaker** [reply]
승록님이 말씀하신 것 처럼, extends의 장점이 있기때문에 컴포넌트 Prop에서는 interface가 더 좋아보입니다!

Todo 구현시에도 다음과 같이 extends를 사용하였구요!

```javascript
export interface TodoType {
  id: number
  todo: string
  isCompleted: boolean
  userId: number
}
```

```javascript
export interface TodoItemProps extends Omit<TodoType, 'userId'> {
  updateTodo: (id: number, todo: string, isCompleted: boolean) => void
  deleteTodo: (id: number) => void
}
```

**Comment by SeungrokYoon** [comment]
페이지 컴포넌트 네이밍 컨벤션을 정하는 것이 어떤지 여러분들의 의견을 듣고 싶습니다!

컴포넌트명+ Props로 타입이 컴포넌트 프롭 관련 타입임을 명시하고 있지만 아직 충분하지 않다고 생각해서요.

일례로, 저는 SigninProps와 SignupProps가 페이지가 아닌 컴포넌트의 프롭타입인줄 착각을 했었습니다.

그 이유는 페이지 컴포넌트인 Signin, Signup과, 폼 컴포넌트인 SignForm 컴포넌트의 네이밍이 비슷하기 때문이었습니다!

그래서 페이지 컴포넌트는 특별히 뒤에 "Page" 접미사를 붙이면 어떨까 해요.

"HomePage", "SigninPage" 이렇게요. 그러면 

`컴포넌트명` + "Props"이라는 타입 컨벤션을  유지하면서도 페이지관련 타입임을 뚜렷이 알 수 있는 긍정적인 효과가 있을 듯 합니다. 

```
interface HomePageProps {
  target:string; ...
}
```

여러분들은 어떻게 생각하시나요?

**Reply by wisdomin121** [reply]
가끔 헷갈릴 때가 있어서 붙이는 거 좋은 것 같습니다 !!

**Reply by 5unk3n** [reply]
저도 페이지와 컴포넌트 이름이 가끔 헷갈려서 Page 접미사 붙이면 좋을 것 같아요!
